### PR TITLE
Reduce API calls and fix NFT display issue

### PR DESCRIPTION
**Pull request by h0ngcha0**
1. Fix the issue where NFTs are not displayed when address has >80 NFTs
2. Do not retry `multi-call` on invalid tokens
3. When loading transaction history in background, if possible, only load the history of the currently selected account instead of accounts from all networks
4. Reduce the call for network liveness check, avoid excessive balance call

### CODE REVIEW

File: packages/extension/src/ui/features/accountTokens/tokens.state.ts [hunk 1]
```diff
@@ -337,17 +338,28 @@ async function fetchFungibleTokenFromFullNode(network: Network, tokenId: string)
...
-      name: Buffer.from(metadata.name, 'hex').toString('utf8'),
-      symbol: Buffer.from(metadata.symbol, 'hex').toString('utf8'),
-      decimals: metadata.decimals,
-      verified: false
+    const metadata = await fetchImmutable(`${tokenId}-token-metadata`, async () => {
+      try {
+        return (await nodeProvider.fetchFungibleTokenMetaData(tokenId))
+      } catch (e: any) {
+        if (e.message.startsWith('Failed to call contract')) {
+          return { name: undefined, symbol: undefined, decimals: 0 }
```
Comment (Lbqds) [inline packages/extension/src/ui/features/accountTokens/tokens.state.ts:346]: [Nit]: Perhaps return `undefined` directly.

**Review by Lbqds — APPROVED** [review]
LGTM

File: packages/extension/src/ui/features/accountTokens/tokens.state.ts (hunk 1 above)
Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountTokens/tokens.state.ts:346]: If we return `undefined` it would be the same as not cached ([here](https://github.com/alephium/extension-wallet/blob/master/packages/extension/src/shared/utils/fetchImmutable.ts#L11-L12)), this way it can be cached so it will not make additional calls.

Comment (Lbqds) [reply packages/extension/src/ui/features/accountTokens/tokens.state.ts:346]: Got it, thanks for the explanation

File: packages/extension/src/ui/features/accountNfts/useNFTCollections.ts [hunk 2]
```diff
@@ -22,6 +22,7 @@ export const useCollectionAndNFTs = (
     () => fetchCollectionAndNfts(nonFungibleTokenIds, network),
     {
       ...config,
+      dedupingInterval: 5000,
```
Comment (polarker) [inline packages/extension/src/ui/features/accountNfts/useNFTCollections.ts:25]: Is this change necessary from 2 seconds to 5 seconds?

In what cases, will the wallet try to query the same data within a few seconds?

Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountNfts/useNFTCollections.ts:25]: When I click around in the extension wallet as a user while monitoring the network tab (e.g. switching tabs back and forth, etc), it feels that 2 seconds is still causing unnecessary calls. 

Not sure what's the best number to set here but feels that 5 seconds might be a good spot between timely update and reducing unnecessary calls.  WDYT?

File: packages/extension/src/ui/features/accountTokens/tokens.state.ts [hunk 3]
```diff
@@ -337,17 +338,28 @@ async function fetchFungibleTokenFromFullNode(network: Network, tokenId: string)
...
+        } else {
+          throw e
+        }
```
[the remark this answers is not in the document]
Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountTokens/tokens.state.ts:349]: It should be okey since this is for fungible tokens.

File: packages/extension/src/background/networkStatus.ts [hunk 4]
```diff
@@ -13,7 +13,7 @@ const swrStorage = new KeyValueStorage<Record<SwrCacheKey, any>>(
 // see: https://github.com/jperasmus/stale-while-revalidate-cache#configuration
 const swr = createStaleWhileRevalidateCache({
   storage: swrStorage, // can be any object with getItem and setItem methods
-  minTimeToStale: 60e3, // 1 minute
+  minTimeToStale: 2 * 60e3, // 2 minutes
   maxTimeToLive: 30 * 60e3, // 30 minutes
 })
```
[the remark this answers is not in the document]
Comment (h0ngcha0) [reply packages/extension/src/background/networkStatus.ts:18]: This will not fire API in the background. The wallet UI queries this periodically when the wallet is open, and `minTimeToStale` determines how long the network status is cached in background before it does the network requests again. Now it is set to 2 minutes instead.

File: packages/extension/src/ui/features/accountTokens/tokens.state.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (polarker) [reply packages/extension/src/ui/features/accountTokens/tokens.state.ts:349]: Will the wallet keep fetching? For example, if there is an assertion error in the code

File: packages/extension/src/ui/features/accountNfts/useNFTCollections.ts (hunk 2 above)
Comment (polarker) [reply packages/extension/src/ui/features/accountNfts/useNFTCollections.ts:25]: If all of the token&NFT data is cached, there should be no API calls. Why there will be extra calls when clicking around?

Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountNfts/useNFTCollections.ts:25]: There are two calls that are not cached, where we have adjusted `dedupingInterval` for:

1. The call to get account balance
2. For NFTs, we have a call to fetch metadata from the explorer backend in batch [here](https://github.com/alephium/extension-wallet/blob/fix-display-nft/packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts#L16-L20), since the payload of the call is dynamic.

File: packages/extension/src/ui/features/accountTokens/tokens.state.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountTokens/tokens.state.ts:349]: For other exceptions it will not be cached, so next time when we display fungible tokens, it will try to make requests again.

We want to avoid the situation where the error is transient and we inadvertently cached the token as invalid, and we know `Failed to call contract` error is persistent. Please let me know if there are better ways to do this.

File: packages/extension/src/ui/features/accountNfts/useNFTCollections.ts (hunk 2 above)
Comment (polarker) [reply packages/extension/src/ui/features/accountNfts/useNFTCollections.ts:25]: So for NFT, we will need to re-fetch metadata every time we click on the NFT screen?

Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountNfts/useNFTCollections.ts:25]: Yeah, we used to cache them when we fetch the NFT metadata individually, after migrating to fetching the metadata in batch from explorer backend, it is not cached any more due to its dynamic payload.

I will try to cache them.

File: packages/extension/src/ui/features/accountTokens/tokens.state.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (polarker) [reply packages/extension/src/ui/features/accountTokens/tokens.state.ts:349]: I think the general practice is to fetch the data using backoff strategies. If the data can't be fetched multiple times, then it's considered unavailable, unless the error is 429. SWR handles 429 differently from other errors in our configuration.

File: packages/extension/src/shared/utils/fetchImmutable.ts [hunk 5]
```diff
@@ -20,6 +20,14 @@ export async function fetchImmutable<Data>(key: string, fetcher: Fetcher<Data, s
   }
 }
 
+export async function getImmutable<Data>(key: string): Promise<Data | undefined> {
+  return await fetchImmutableStorage.get(key)
+}
+
+export async function storeImmutable<Data>(key: string, data: Data): Promise<void> {
+  return await fetchImmutableStorage.set(key, data)
+}
```
Comment (polarker) [inline packages/extension/src/shared/utils/fetchImmutable.ts:29]: Local storage has size constraints, so if the cache grows too large, it can consume all available local storage.

This is something to be aware of, although we can improve it later.

File: packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts [hunk 6]
```diff
@@ -103,11 +103,14 @@ async function getNftMetadataz(nftIds: string[], explorerProvider: ExplorerProvi
...
+  if (nftIdsWithoutMetadata.length !== 0) {
+    newNftMetadataz = (await Promise.all(
+      chunk(nftIdsWithoutMetadata, maxSizeTokens).map((ids) =>
+        explorerProvider.tokens.postTokensNftMetadata(ids)
+      )
+    )).flat()
+  }
 
   for (const newNftMetadata of newNftMetadataz) {
     await storeImmutable(`${newNftMetadata.id}-nft-metadata`, newNftMetadata)
```
Comment (polarker) [inline packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts:116]: 1. Will it be a problem that we create too many keys in the local storage?
2. Maybe it's better to use prefixes than postfixes?

File: packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts [hunk 7]
```diff
@@ -103,11 +103,14 @@ async function getNftMetadataz(nftIds: string[], explorerProvider: ExplorerProvi
...
-      explorerProvider.tokens.postTokensNftMetadata(ids)
-    )
-  )).flat()
+  let newNftMetadataz: NFTMetadata[] = []
+  if (nftIdsWithoutMetadata.length !== 0) {
+    newNftMetadataz = (await Promise.all(
+      chunk(nftIdsWithoutMetadata, maxSizeTokens).map((ids) =>
+        explorerProvider.tokens.postTokensNftMetadata(ids)
+      )
+    )).flat()
```
Comment (polarker) [inline packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts:112]: If one of the calls fails, does it mean that none of the successful calls will be cached?

File: packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts (hunk 6 above)
Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts:116]: > Will it be a problem that we create too many keys in the local storage?

There isn't a limit to the number of keys, but local storage dose have a size constrains (10MB) which will put an upper limit to that. For token metadata, each entry is around 0.2~0.3kb, and we would need around a low tens of thousands of tokens to exceed the size limit (together with other things such as tx history). 
From what I read the individual key access is fast.

> Maybe it's better to use prefixes than postfixes?

I don't have preference here. If there are good benefits with postfixes I will update. (seems that it won't affect performance)

File: packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts (hunk 7 above)
Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts:112]: Great point, i will update.

File: packages/extension/src/ui/features/accountTokens/tokens.state.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountTokens/tokens.state.ts:349]: Yes, exceptions will be subject to the SWR retry strategy. For fetching fungible token metadata specifically, [current SWR setup](https://github.com/alephium/extension-wallet/blob/master/packages/extension/src/ui/features/accountTokens/tokens.state.ts#L207-L212) is to retry if the error is 429 using SWR's default retry strategy, and not retry for other types of exceptions.

In this case, we are handling a known and persistent exception for getting token metadata and cache it as invalid, such that when user call it again it wont make more requests, I guess this is on top of the SWR retry strategy. WDTY?

File: packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts (hunk 6 above)
Comment (polarker) [reply packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts]: Prefix is more readable and easier to sort keys based on key types.

File: packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts [hunk 8]
```diff
@@ -86,3 +89,30 @@ async function getCollectionMetadata(
   const metadataResponse = await fetch(metadata.collectionUri)
   return await metadataResponse.json()
 }
+
+async function getNftMetadataz(nftIds: string[], explorerProvider: ExplorerProvider) {
```
Comment (polarker) [inline packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts:93]: Does NFT metadata fetch use backoff retry when the query fails?

Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts:93]: Yes, it uses SWR's default exponential backoff retry strategy when query fails.

File: packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts (hunk 6 above)
Comment (h0ngcha0) [reply packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts]: I will update

File: packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts (hunk 8 above)
Comment (polarker) [reply packages/extension/src/ui/features/accountNfts/alephium-nft.service.ts:93]: Yes. Should be consistent

File: packages/extension/src/ui/services/swr.ts [hunk 9]
```diff
@@ -104,3 +104,8 @@ export const swrCacheProvider: Cache = {
     }
   },
 }
+
+export const retryWhenRateLimited = (error: any) => {
+  const errorCode = error?.status || error?.errorCode
```
Comment (polarker) [inline packages/extension/src/ui/services/swr.ts:109]: Our API error does not have a status or error code. This may not work?

Comment (h0ngcha0) [reply packages/extension/src/ui/services/swr.ts]: Indeed, you are right. 

I have upgraded the SDK version so we can [extract](https://github.com/alephium/extension-wallet/pull/212/commits/ce77bab18a1e86e393f515907974fb55f6c61a2c) the error code. Maybe we should introduce a util function to extract error code in SDK instead?
